config/helpers/bootstrap.go

```diff
@@ -0,0 +1,193 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package helpers
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"embed"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	extensionsapiserver "k8s.io/apiextensions-apiserver/pkg/apiserver"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	apimachineryerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/klog/v2"
+
+	configcrds "github.com/kcp-dev/kcp/config/crds"
+)
+
+// TransformFileFunc transforms a resource file before being applied to the cluster.
+type TransformFileFunc func(bs []byte) ([]byte, error)
+
+// Option allows to customize the bootstrap process.
+type Option struct {
+	// TransformFileFunc is a function that transforms a resource file before being applied to the cluster.
+	TransformFile TransformFileFunc
+}
+
+// ReplaceOption allows to customize the bootstrap process.
+func ReplaceOption(pairs ...string) Option {
+	return Option{
+		TransformFile: func(bs []byte) ([]byte, error) {
+			if len(pairs)%2 != 0 {
+				return nil, fmt.Errorf("odd number of arguments: %v", pairs)
+			}
+			for i := 0; i < len(pairs); i += 2 {
+				bs = bytes.ReplaceAll(bs, []byte(pairs[i]), []byte(pairs[i+1]))
+			}
+			return bs, nil
+		},
+	}
+}
+
+// Bootstrap creates a list of CRDs and then the resources in a package's fs by
+// continuously retrying the list. This is blocking, i.e. it only returns (with error)
+// when the context is closed or with nil when the bootstrapping is successfully completed.
+func Bootstrap(ctx context.Context, crdClient apiextensionsclient.Interface, dynamicClient dynamic.Interface, fs embed.FS, crds []metav1.GroupResource, opts ...Option) error {
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(crdClient.Discovery()))
+
+	if err := wait.PollImmediateInfiniteWithContext(ctx, time.Second, func(ctx context.Context) (bool, error) {
+		if err := configcrds.Create(ctx, crdClient.ApiextensionsV1().CustomResourceDefinitions(), crds...); err != nil {
+			klog.Errorf("failed to bootstrap CRDs: %v", err)
+			return false, nil // keep retrying
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("failed to bootstrap CRDs into root workspace: %w", err)
+	}
+
+	// bootstrap non-crd resources
```

Comment:
What's this for?

Reply:
E.g. workspace types, or any other kind of resources we need like roles/rolebindings.

Reply:
Ok, so mostly forward-looking at this point?

Reply:
yes